java/arcs/core/data/Annotation.kt

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.data
+
+/** A class containing an annotation. */
+data class Annotation(val name: String, val params: Map<String, AnnotationParam>) {
+    companion object {
+        private val ANNOTATION_PATTERN = "^([a-zA-Z]+)(\\((.+)\\))?$".toRegex()
+        private val PARAM_PATTERN = "^([a-zA-Z]+):(.*)$".toRegex()
+
+        fun fromString(str: String): List<Annotation> {
+            val tokens = str.split("@")
+            return tokens.drop(1).map {
+                val match = requireNotNull(ANNOTATION_PATTERN.matchEntire(it.trim())) {
+                    "invalid annotation $it"
+                }
+                val params = mutableMapOf<String, AnnotationParam>()
+                if (match.groupValues[3].length > 0) {
```

Comment:
These groups are easier to read if you use destructuring to assign them to variables:

```
val (_, name, _, params) = match.groupValues
```

(I'm not actually sure what each element in the list represents, but you can use _ to ignore ones you don't care about)

Reply:
Additionally, use `.isNotEmpty` instead of `.length > 0`

Reply:
that's cool, updated the code thanks!